### PR TITLE
Consistently use 64-bit integers in `io.ascii` (including Windows)

### LINKED ISSUE #5744: Windows ascii (and Table) fails on integer data that can't be represented as int32 (but could be as int64)
For example:

```
ascii.read("""num,ra,dec,radius,mag
100000000000000000,32.23222,10.1211,0.8,18.1
2,38.12321,-88.1321,2.2,17.0
""")
```

gives the following warnings:

```
WARNING: OverflowError converting to IntType in column num, reverting to String. [astropy.io.ascii.fastbasic]
WARNING:astropy:OverflowError converting to IntType in column num, reverting to String.
```

However the first line could be easily interpreted as `int64`:

```
ascii.read("""num,ra,dec,radius,mag
100000000000000000,32.23222,10.1211,0.8,18.1
2,38.12321,-88.1321,2.2,17.0
""")['num'].astype(np.int64)
```

Could the default integer type (I know that's a mess on windows) be set to int64?

### PR DESCRIPTION
**Pull request by orionlee**
### Description

This pull request is to change `io.ascii` to always use 64-bit integers when reading integer numeric table data. Previously the length of the integer field (32-bit or 64-bit) was system-dependent, with Windows being a notable example of defaulting to 32-bit integer. This caused long integer numeric (which overflowed 32-bits) to be returned as strings on Windows.

Fixes #5744

This also address https://github.com/astropy/astropy/issues/15989 for `main` (v6.1 and later). The fix for v6.0.x series for this issue is at https://github.com/astropy/astropy/pull/15992

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Comment by pllim** [comment]
Thanks! I am guessing this also supersedes https://github.com/astropy/astropy/pull/15992 and fixes https://github.com/astropy/astropy/issues/15989 as well?

When you think it is ready, I will be happy to help you ping reviewers.

**Comment by dhomeier** [comment]
Thanks, marking this for extra CI to check `long long` on other platforms (though I honestly have no idea if they are > 64bit on any of those).

**Comment by orionlee** [comment]
#pilim This fixes #15989 as well.

For PR #15992, mhvk suggested merging it anyway. I'm okay either way, it is defensive in nature and can possibly guard against any bug / behavioral changes of default conversion.

One compromise would be folding the ipac-specific test in 15992 into this PR, such that there is at least a test to guard against possible future regression.

**Comment by pllim** [comment]
@mhvk , if this is good to proceed, then no need to merge https://github.com/astropy/astropy/pull/15992 ?

**Comment by mhvk** [comment]
> @mhvk , if this is good to proceed, then no need to merge #15992 ?

This looks pretty good to me! But if we're not backporting, then perhaps #15992 should still be done for 6.0.x?

**Comment by pllim** [comment]
We can consider backporting this if you think it won't be "breaking change". What do you think?

**Comment by taldcroft** [comment]
Thanks @orionlee for finally having the courage to address this years-old issue.

I believe this PR needs to be flagged as an API change, which answers any question about backporting. Some users might be depending on the output being int32, and others might be depending on long ints being parsed as strings. The change log text might read:
```
The ``io.ascii`` Python and C table readers were updated to always use a 64-bit integer field when
reading a column of integer numeric data. This changes the default behavior on Windows and
potentially 32-bit architectures. Previously on those platforms, table columns with any long 
integers which overflowed the 32-bit integer would be returned as a string column.
```
I'm not sure if 32-bit linux would previously give 32-bit or 64-bit int, but this wording just gives a
heads-up that things might change.

**Comment by taldcroft** [comment]
I've taken the liberty to update the description and PR title to reflect what I see as the big-picture change, namely changing the default integer to always be 64-bits. This is a potentially breaking change for some users so it needs to be highlighted.

**Comment by taldcroft** [comment]
Am I am understanding the code correctly that this is the case? It seems clear for the Python reader, but I'm not 100% sure on the C-reader.

I see some test code that indicates the default precision could end up being more than 64-bits (`assert dat["num"].dtype.itemsize >= 8`). I would rather see the tests asserting `itemsize == 8` if that is indeed the expectation.

**Comment by mhvk** [comment]
For the changelog, we can perhaps add that this follows numpy >= 2.0 behaviour - to make clear this is not done in isolation.

**Comment by pllim** [comment]
Re: https://github.com/astropy/astropy/pull/16005#issuecomment-1933231312

OK, so here is the plan, I am going to merge https://github.com/astropy/astropy/pull/15992 then and have that backported. But then this PR will have to rebase and revert that patch but only for `main`? Am I understanding that correctly?

If we're reverting in `main` anyway, should #15992 just be opened directly against v6.0.x branch instead of `main`?

**Comment by mhvk** [comment]
> If we're reverting in `main` anyway, should #15992 just be opened directly against v6.0.x branch instead of `main`?

That seems easiest!

**Comment by pllim** [comment]
@orionlee , would you be willing to redo https://github.com/astropy/astropy/pull/15992 so it goes straight to v6.0.x branch instead of main ? Usually by checking out v6.0.x branch and then cherry picking from your patch, that would be enough. If not, let me know and I will be happy to do it, but in that case git would mistaken me as the author and the best I can do is credit you as co-authore. Please let me know how you want to proceed. Thanks!

**Comment by eerovaher** [comment]
> ... but in that case git would mistaken me as the author and the best I can do is credit you as co-author.

Git associates with every commit its author and its committer and they can be different people. See https://git-scm.com/book/en/v2/Git-Basics-Viewing-the-Commit-History and in particular the paragraph just after Table 1.

**Comment by neutrinoceros** [comment]
> in that case git would mistaken me as the author and the best I can do is credit you as co-author.

@pllim actually `git cherry-pick` correctly preserves authorship ;)

**Comment by pllim** [comment]
Does it? 🤔  But what if there is conflict? I guess I can try if @orionlee does not want to.

**Comment by dhomeier** [comment]
Thanks, C reader looks good to me, too, and I made some tests with the `ascii_read_bench` notebook to check that performance is not affected (on macOS, so the only real difference is probably using `strtoi64` around `strtol[l]`).

Out of curiosity, have you also considered using `sscanf` as suggested in the [second stackoverflow answer](https://stackoverflow.com/a/17003732)? It looks like it might be a bit more direct, but I am not sure if the error handling is as adequate.

File: docs/changes/io.ascii/16005.bugfix.rst [hunk 1]
```diff
@@ -0,0 +1,2 @@
+Fixed reading large (64 bits) integers on some platforms, e.g.,  Windows.
+Such columns are no longer read as String.
```
Comment (pllim) [inline docs/changes/io.ascii/16005.bugfix.rst:2]:
```suggestion
Such columns are no longer read as String. This includes reading IPAC tables for ``long`` column type.
```

**Comment by pllim** [comment]
@eerovaher , the `astropy/coordinates/tests/test_angles.py` failures in devinfra job is unrelated, right?

File: docs/changes/io.ascii/16005.bugfix.rst [hunk 2]
```diff
@@ -0,0 +1,5 @@
+The ``io.ascii`` Python and C table readers were updated to use a 64-bit integer field by
+default when reading a column of integer numeric data. This changes the default behavior
+on Windows and potentially 32-bit architectures. Previously on those platforms, table
+columns with any long  integers which overflowed the 32-bit integer would be returned
+as a string column. The new default behavior is consistent with `numpy` v2 and `pandas`.
```
Comment (pllim) [inline docs/changes/io.ascii/16005.bugfix.rst:5]: We do not do "live linking" in change log because we do not want to go back in time to fix broken links from the past.

```suggestion
as a string column. The new default behavior is consistent with ``numpy`` v2 and ``pandas``.
```

Also, do you want to mention IPAC here?

**Comment by orionlee** [comment]
@taldcroft The test has been updated to assert exactly 64 bit integers. It also explicitly tested 64 bit integers are used even for columns with only small values.
(In an earlier iteration, C codes were implemented with `long long` instead of `int64_t`. So I wasn't sure if it might be 128 bit in some cases. )

The latest changelog:

```rst
The ``io.ascii`` Python and C table readers were updated to use a 64-bit integer field by
default when reading a column of integer numeric data. This changes the default behavior
on Windows and potentially 32-bit architectures. Previously on those platforms, table
columns with any long  integers which overflowed the 32-bit integer would be returned
as a string column. The new default behavior is consistent with `numpy` v2 and `pandas`.
```

File: docs/changes/io.ascii/16005.bugfix.rst (hunk 2 above)
Comment (orionlee) [reply docs/changes/io.ascii/16005.bugfix.rst]: See the latest revison.

**Comment by orionlee** [comment]
One questions to the experts: are there some other supported, non-ascii formats that have similar issues, and should be updated (to use `np.int64` for integer columns by default) as well?

Presumably the issue is not applicable for formats that require explicit typing like FITS, I am not sure if that's the case for all non-ascii formats.

**Comment by orionlee** [comment]
@dhomeier On the details of C implementation, I have not looked too closely on the `sscanf` approach. I picked `strtoll()` as the basis primarily because I think it's easier to maintain, as the abstraction / API is consistent with the existing codes for `double`.

**Comment by dhomeier** [comment]
I agree keeping a consistent API is better here; I just considered `sscanf` as it might be used directly with `SCNd64` so no check for the range of `long long` would be needed, but on the down side it seems to automatically truncate all inputs without any obvious way to raise an overflow error (+ conversion errors would need to be handled completely differently).
And on top, `sscanf` is noticeably slower, while on a platform where `int64` is already default, `strtoll` is keeping conversion speed unchanged; for reference Numpy's new C parser is now generally the fastest option for integers:
read_strtoll
read_sscanf

That perhaps to keep #12829 in mind now that our minimum Numpy version is 1.23; for floating-point conversions however our parsers are still noticeably faster.

**Comment by orionlee** [comment]
@dhomeier Currently in the PR, the C code path does not perform any range check on `long long` at runtime, unless `long long` is larger than 64 bits (through the use of macros). It is essentially `(int64_t)strtoll()` wrapped in a function. 

If we really want, it could be further inlined using macros: 
```c
#if LLONG_MIN >= INT64_MIN && LLONG_MAX <= INT64_MAX
    #define strtoi64(nptr, endptr, base) (int64_t)strtoll(nptr, endptr, base)
#else
... // the function form with range check for case long long > 64 bits
```

I thought it might be a tad too clever, but it looks like macros have been used quite a bit in `tokenizer.c` already.

Now I think about it, it might be best to change the new `strtoi64()` function to `static inline`. The effect is probably (almost) the same as the inline macros above.

**Review by dhomeier — APPROVED** [review]
> @dhomeier Currently in the PR, the C code path does not perform any range check on `long long` at runtime, unless `long long` is larger than 64 bits (through the use of macros). It is essentially `(int64_t)strtoll()` wrapped in a function.

In the other cases `strtoll` will already raise `ERANGE` afaics; otherwise it would fail `test_convert_overflow` (`sscanf` does not). So it _should_ work the same with the range check for `LLONG_MIN < INT64_MIN` here, though I guess we don't test on any architectures where that branch applies.

`static inline` function as now should be fine, none of the wrapper code seems to have any performance impact (I was wondering if the `(int64_t)` cast might be moved inside the `#if LLONG_MIN` as well, but seems to make no timing difference either).

**Comment by dhomeier** [comment]
BTW this might also replace / resolve #13359?

**Comment by orionlee** [comment]
It looks to me this PR does replace and resolve #13359 . 

#13359 implements the behavior of "if default int conversion fails due to overflow, try int64" in Python code path, the option 2 in https://github.com/astropy/astropy/issues/5744#issuecomment-1930501674 . This PR implements options, which use int64 by default.

**Comment by orionlee** [comment]
One minor issue: for IPAC tables `int` columns, after this PR, it's going to be `int64`.

Specifically, the behavior for `int` columns is:

|                | `stable` | PR    |
| -------------- | ------ | ----- |
| Windows        | *int32*  | int64 |
| Most platforms | int64  | int64 |

The PR brings parity between Windows and most platforms. But arguably `int` columns should be `int32` in the first place. Thoughts?

**Comment by mhvk** [comment]
Agreed that, in principle, if a file specifically asks for `int`, it should be `int32`. Though given that that is for ASCII data, and that we have not had any complaints, it is not a big issue... (while it is a *bug* to give int32 when one asks for int64!)

**Comment by orionlee** [comment]
[IPAC Spec](https://irsa.ipac.caltech.edu/applications/DDGEN/Doc/ipac_tbl.html#datatypes) does not define the range of `int` / `long` .

Leaving it alone is probably the most pragmatic choice, as there won't be potentially breaking behavior.

An edge case is that round trips involving IPAC tables would lead to `int` columns become `long` in export. It's an existing issue with no complaint though. E.g.,

```python
import astropy.io.ascii as ascii

test_data = """\
|              oid|expid|cadence|
|             long|    l|    int|
 90000000000000001   123       1
"""
dat = ascii.read(test_data, format="ipac")

ascii.write(dat, "ipac_export.tbl", format="ipac")
```

The `cadence` column's type is changed from `int` to `long` in the exported `ipac_export.tbl` file.

**Comment by taldcroft** [comment]
@dhomeier - thanks for making the great performance comparison charts, very useful! My takeaway is that pandas and `np.loadtxt` are about 50% faster, which is not trivial but not horrible. The existing C reader has the advantage of being strictly compliant with the behavior of the Python reader. There are some subtleties in the parsing that will show up if/when we try to use `np.loadtxt`. At least for the moment we can keep that idea in mind but there is not a huge driver.

Otherwise I think the interface impacts of consistently using int64 always are acceptable and the uniformity across platforms will be welcome by most people. Thanks to all for the very careful work on this one!

**Review by taldcroft — APPROVED** [review]
Looks good to me, with the explicit caveat that I have not reviewed the details of the `static inline int64_t strtoi64` code. This has been reviewed by @dhomeier.

**Comment by taldcroft** [comment]
I approved this, but I do think this should get a What's New entry. @orionlee - can you add that? (I see now that adding the Whats-new-needed label has kicked off a whole new CI run, ugh).

Any last thoughts?

**Comment by taldcroft** [comment]
Add this point if there are no further code changes can we turn off all the cron testing?

**Comment by dhomeier** [comment]
Agreed, don't need it for doc changes (and I had no idea that adding a label would trigger them either)!

**Comment by dhomeier** [comment]
> My takeaway is that pandas and `np.loadtxt` are about 50% faster, which is not trivial but not horrible. The existing C reader has the advantage of being strictly compliant with the behavior of the Python reader. There are some subtleties in the parsing that will show up if/when we try to use `np.loadtxt`. At least for the moment we can keep that idea in mind but there is not a huge driver.

The actual numbers are closer to 30-35% for large tables, but from the discussion in #12829 emerged that `np.loadtxt` does not support e.g. all the comment settings that `io.ascii` does (though not all the formats are supported by our fast reader either). Dropping it in somehow at the top level would thus not be straightforward, so I've been wondering if we can plugin the numpy parser at a lower level like `cparser.Cparser`. But it is probably not clear if that would still retain (all) the performance advantage from `loadtxt`.

**Comment by orionlee** [comment]
@taldcroft A What's New entry has been added.

**Comment by pllim** [comment]
Given 2 approvals and detailed discussions, I merged this. Thank you, all!

**Comment by orionlee** [comment]
@taldcroft @pllim Should the changelog entry `16005.bugfix.rst` be named as `16005.feature.rst` instead, given this PR is not going to backported?

I don't know if the filename might trigger some kind of (not applicable) workflow down the road.

**Comment by taldcroft** [comment]
@orionlee - good catch and sorry I missed that in review. I actually think this should be `16005.api.rst` for the API change. This can be done as a follow-on PR to just change the file name.

**Comment by dhomeier** [comment]
I think it would have triggered the towncrier check if it _had_ been named differently, as this is still labelled `Bug`, though admittedly this is a borderline case, also involving an `API change` – but it does resolve a buggy behaviour.

**Comment by dhomeier** [comment]
> To me, it is still a bugfix, but just one we cannot backport because it is also a bit of an API change. So perhaps it should have been a `16005.api.rst` instead... But really that is also covered by your what's-new entry, and numpy >= 2.0 gives it to us as well. Since this also is a bugfix, I think it is OK...

I also think the changelog entry is clear enough, but the standard procedure would probably have been to add it both in `bugfix` and `api` – arguably documenting the API change here is important.

**Comment by taldcroft** [comment]
@mhvk - I think we often have a different view of bugs. I view #5744 as not a bug (and in fact it does not have a `bug` label). The behavior of astropy prior to this PR was consistent with design expectations, namely that it used the default `int` size for the system. It happened that a common use case (catalog IDs that exceed 32-bits but not 64-bits) showed that this behavior is **undesirable** and we changed it, but it was never a bug in the code. The code was behaving exactly as designed and expected.

So I call something a bug if the code is returning an incorrect value relative to the design expectation and documentation. It is important to note the distinction between "design expectation" and "user expectation" (which is sometimes influenced by documentation). 

Remember that even after this fix, columns with integers that exceed 64-bits will *still* be returned as strings.

**Comment by mhvk** [comment]
Fair enough! I think I partially had #15989 in my mind, which this PR fixed too (for 6.1), and which was a proper bug, since the IPAC table explicitly asked for long.

**Comment by pllim** [comment]
So, in conclusion, is https://github.com/astropy/astropy/pull/16026 needed? If so, is it good to merge?

**Comment by dhomeier** [comment]
> Fair enough! I think I partially had #15989 in my mind, which this PR fixed too (for 6.1), and which was a proper bug, since the IPAC table explicitly asked for long.

But here we have rather the opposite case, as IPAC `int` and `long` columns are not treated as `int64` alike, and according to https://github.com/astropy/astropy/pull/16005#issuecomment-1936653697 will be written back as `long` when exported again.
I agree that the original issue of some ASCII input numbers overflowing the format is not a bug; we've discussed that before, and there just is not the _right_ dtype to fit everything. To me it's rather ensuring consistent behaviour across all (currently know or supported) platforms that could be considered a bugfix, but if it is not, I'm fine with that as well.

Bottom line: we certainly need #16026 to document it as API change, the question is perhaps rather if we remove it from the `Bug` section after all.

**Comment by orionlee** [comment]
How about specifying the behavior change in API section, and specifying the IPAC table `long` column fix in bug section?

**Comment by mhvk** [comment]
> How about specifying the behavior change in API section, and specifying the IPAC table `long` column fix in bug section?

Yes, makes sense!

**Comment by orionlee** [comment]
Also, for behavior of integer columns exceeding `np.int64` . `pandas` would attempt to use `np.uint64`. If the number is larger than `np.uint64`, it then use String.

Test script
```
from io import StringIO
import numpy as np
import pandas as pd

print('numpy ', np.__version__)
print('pandas ', pd.__version__)

print(np.iinfo(np.int64).max + 1, np.iinfo(np.uint64).max + 1)

df = pd.read_csv(StringIO("""\
col0,col1,col2
1,9223372036854775808,18446744073709551616
2,23,45
"""))

# print(df)
print([df[c].dtype for c in df.columns])
```

Result:
```python
numpy  1.26.3
pandas  2.1.4
9223372036854775808 18446744073709551616
[dtype('int64'), dtype('uint64'), dtype('O')]
```

The behavior is the same on pandas 1.5.3
